controllers/helpers.go

```diff
@@ -113,3 +122,207 @@ func GetObjectsToRequestsByNamespaceAndClusterName(ctx context.Context, c client
 	}
 	return results
 }
+
+// Returns true if a and b point to the same object
+func referSameObject(a, b metav1.OwnerReference) bool {
+	aGV, err := schema.ParseGroupVersion(a.APIVersion)
+	if err != nil {
+		return false
+	}
+
+	bGV, err := schema.ParseGroupVersion(b.APIVersion)
+	if err != nil {
+		return false
+	}
+
+	return aGV.Group == bGV.Group && a.Kind == b.Kind && a.Name == b.Name
+}
+
+// GetCloudProviderSecret returns the required azure json secret for the provided parameters. It differentiates between
+func GetCloudProviderSecret(d azure.ClusterDescriber, namespace, name string, identityType infrav1.VMIdentity, userIdentityID string) (*corev1.Secret, error) {
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      fmt.Sprintf("%s-azure-json", name),
+			Labels: map[string]string{
+				string(infrav1.ResourceLifecycleOwned): "",
+			},
+		},
+	}
+
+	var data []byte
+	var err error
+
+	switch identityType {
+	case infrav1.VMIdentitySystemAssigned:
+		data, err = systemMSICloudProviderConfig(d)
+		if err != nil {
+			return nil, err
+		}
+	case infrav1.VMIdentityUserAssigned:
+		if len(userIdentityID) < 1 {
+			return nil, errors.New("expected a non-empty userIdentityID")
+		}
+		data, err = userMSICloudProviderConfig(d, userIdentityID)
+		if err != nil {
+			return nil, err
+		}
+	case infrav1.VMIdentityNone:
+		data, err = servicePrincipalCloudProviderConfig(d)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	secret.Data = map[string][]byte{
+		"azure.json": data,
+	}
+
+	return secret, nil
+}
+
+func servicePrincipalCloudProviderConfig(d azure.ClusterDescriber) ([]byte, error) {
+	return json.MarshalIndent(newCloudProviderConfig(d), "", "    ")
+}
+
+func systemMSICloudProviderConfig(d azure.ClusterDescriber) ([]byte, error) {
```

Comment:
I think MSI is old naming for these. What do you think about sticking with SystemAssigned and UserAssigned Managed Identities.